tests/Application/Actions/Donations/UpdateTest.php

```diff
@@ -1218,6 +1219,210 @@ public function testAddDataHitsAlreadyCapturedStripeExceptionWithFeeChange(): vo
+        $donationRepoProphecy
+            ->deriveFees(Argument::type(Donation::class), null, null)
+            ->willReturn($donation) // Actual fee calculation is tested elsewhere.
+            ->shouldBeCalledOnce();
+
+        // Internal persist still goes ahead.
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+        $entityManagerProphecy->beginTransaction()->shouldBeCalledOnce();
+        $entityManagerProphecy->rollback()->shouldBeCalledOnce();
+
+        $stripePaymentIntentsProphecy = $this->prophesize(PaymentIntentService::class);
+
+        $mockPI = new PaymentIntent();
+        $mockPI->application_fee_amount = 526;
+
+        $stripePaymentIntentsProphecy->update('pi_externalId_123', [
+            'amount' => 12_666,
+            'currency' => 'usd',
+            'metadata' => [
+                'coreDonationGiftAid' => true,
+                'feeCoverAmount' => '0.00',
+                'matchedAmount' => '0.0',
+                'optInCharityEmail' => false,
+                'optInTbgEmail' => true,
+                'salesforceId' => 'sfDonation369',
+                'stripeFeeRechargeGross' => '2.05',
+                'stripeFeeRechargeNet' => '2.05',
+                'stripeFeeRechargeVat' => '0.00',
+                'tbgTipGiftAid' => false,
+                'tipAmount' => '3.21',
+            ],
+            'application_fee_amount' => 526,
+        ])
+            ->shouldBeCalledTimes(2)
+            ->will(new PaymentIntentUpdateAttemptTwicePromise(
+                false,
+                $this->prophesize(PaymentIntent::class)->reveal(),
+            ));
+        $stripeClientProphecy = $this->prophesize(StripeClient::class);
+        $stripeClientProphecy->paymentIntents = $stripePaymentIntentsProphecy->reveal();
+
+        $container->set(DonationRepository::class, $donationRepoProphecy->reveal());
+        $container->set(EntityManagerInterface::class, $entityManagerProphecy->reveal());
+        $container->set(StripeClient::class, $stripeClientProphecy->reveal());
+
+        $request = $this->createRequest(
+            'PUT',
+            '/v1/donations/12345678-1234-1234-1234-1234567890ab',
+            json_encode($donation->toApiModel()),
+        )
+            ->withHeader('x-tbg-auth', DonationToken::create('12345678-1234-1234-1234-1234567890ab'));
+        $route = $this->getRouteWithDonationId('put', '12345678-1234-1234-1234-1234567890ab');
+
+        $response = $app->handle($request->withAttribute('route', $route));
+        $payload = (string) $response->getBody();
+
+        $this->assertJson($payload);
+        $this->assertEquals(500, $response->getStatusCode());
+
+        $payloadArray = json_decode($payload, true);
+        $this->assertEquals([
+            'error' => [
+                'type' => 'SERVER_ERROR',
+                'description' => 'Could not update Stripe Payment Intent [C]',
+            ],
+        ], $payloadArray);
+    }
```

Comment:
at 100 lines this is IMHO quite long and hard to follow as a test. Partly that's because I'm new to the codebase, partly I think it is awkwardly long. Could do some extract function refactorings as an easy way to get it a bit more compact. Same applies to previous test.

Also looks like there's probably quite a bit of code in common between the two tests, so factoring that out might make it easier to see what the differences are.

Reply:
Agree a bit of duplication is generally much better than a bit of unclear abstraction. I'm not really bothered by duplication per-se, just by a test method being too long and complex to easily understand as a whole.